File: android/app/src/main/java/dev/bennett/codexmeter/CodexModelCatalog.java
package dev.bennett.codexmeter;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/** Models visible in the account's Codex picker; usage buckets are not a catalog. */
public final class CodexModelCatalog {
    public static final class Model {
        public final String id;
        public final String name;

        Model(String id, String name) {
            this.id = id;
            this.name = name;
        }
    }

    private CodexModelCatalog() { }

    public static List<Model> parse(String body) throws JSONException {
        JSONArray entries = new JSONObject(body).getJSONArray("models");
        List<Model> models = new ArrayList<>();
        Set<String> seen = new HashSet<>();
        for (int i = 0; i < entries.length(); i++) {
            JSONObject entry = entries.getJSONObject(i);
            String id = entry.getString("slug").trim();
            String visibility = entry.getString("visibility");
            if (!"list".equals(visibility) || id.isEmpty() || !seen.add(id)) continue;
            String name = entry.optString("display_name", "").trim();
            models.add(new Model(id, name.isEmpty() ? id : name));
        }
        return models;
    }

    public static List<Model> additions(Set<String> known, List<Model> models) {
        List<Model> added = new ArrayList<>();
        if (known.isEmpty()) return added; // First nonempty catalog establishes a silent baseline.
        for (Model model : models) if (!known.contains(model.id)) added.add(model);
        return added;
    }
}

File: ios/ModelsMeterCore/Sources/ModelsMeterCore/CodexModelCatalog.swift
import Foundation

public struct CodexModel: Codable, Equatable, Sendable, Identifiable {
    public let id: String
    public let name: String

    public init(id: String, name: String) {
        self.id = id
        self.name = name
    }
}

/// The account's Codex picker catalog, not the public API model list or usage buckets.
public enum CodexModelCatalog {
    public static func parse(_ data: Data) throws -> [CodexModel] {
        struct Response: Decodable { let models: [Entry] }
        struct Entry: Decodable {
            let slug: String
            let display_name: String?
            let visibility: String
        }
        let response = try JSONDecoder().decode(Response.self, from: data)
        var seen = Set<String>()
        return response.models.compactMap { entry in
            let id = entry.slug.trimmingCharacters(in: .whitespacesAndNewlines)
            guard entry.visibility == "list", !id.isEmpty, seen.insert(id).inserted else { return nil }
            let name = entry.display_name?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return CodexModel(id: id, name: name.isEmpty ? id : name)
        }
    }
}

/// Keep every observed ID so removal/reappearance and display-name changes never alert again.
public struct CodexModelDiscoveryState: Codable, Equatable, Sendable {
    public private(set) var knownIDs: Set<String>

    public init(knownIDs: Set<String> = []) { self.knownIDs = knownIDs }

    public func additions(in models: [CodexModel]) -> [CodexModel] {
        guard !knownIDs.isEmpty else { return [] }
        return models.filter { !knownIDs.contains($0.id) }
    }

    public mutating func record(_ models: [CodexModel]) {
        knownIDs.formUnion(models.map(\.id))
    }
}
